### ISSUE TITLE
How to create PCA model?

### ISSUE THREAD

**Issue by samsgates** [patrikhuber/eos#26]
Hi,
Please guide me to create PCA model from .ply or .obj file. I have full head model in obj format. I want to create PCA model from normal 3d model to implement into face-fit module.

**Comment by samsgates** [comment]
The below data list required to implement PCA Model, i have only "triangle_list" (vertices list), let me know how to generate other values

mean - Not available	
pca_basis - Not available	
eigenvalues - Not available	
triangle_list - vertices list from model

send me any sample code (c++) to generate these values

**Comment by patrikhuber** [comment]
Hi,

I would suggest you read up on how PCA models work. Our [VISAPP 2016](http://www.patrikhuber.ch/files/3DMM_Framework_VISAPP_2016.pdf) paper would be a good starting point, alternatively the original [Blanz & Vetter 1999](http://gravis.cs.unibas.ch/Sigg99.html) paper is also a very good read.

The papers will explain mean, PCA basis and eigenvalues.

**Comment by samsgates** [comment]
Shall i use the below code to get mean, PCA basis and eigenvalues

#include <opencv/cv.h>

cv::Mat pt_mat;

cv::PCA pt_pca(pt_mat, cv::Mat(), CV_PCA_DATA_AS_ROW, 0);

// Mean
cv::Mat pt_mean = pt_pca.mean;

// Eigen values
cv::Mat pt_eig_vals = pt_pca.eigenvalues;

// Eigen vectors
cv::Mat pt_eig_vecs = pt_pca.eigenvectors;

**Comment by samsgates** [comment]
Query about MorphableModel from PCA model

MorphableModel(shapeModel, colorModel, texCoords);

shapeModel - i can use normal PCA model
colorModel - how to get color PCA model?
texCoords - is that "vt" values form model?

**Comment by patrikhuber** [comment]
The code looks more or less ok. Applying PCA is not the hard part, it's getting the registered data :-)

`shape_model` is a PCA model over vertex positions. `colour_model` is the model over the RGB vertex colours. `tex_coords` is I believe optional. It's the uv-coordinates for each vertex, used for texture mapping.

**Comment by samsgates** [comment]
i have created PCA model and load into eos PCA as below, but vertices count reutrun as "0" in mesh module

 eos::morphablemodel::PcaModel shapeModel(pt_mean, pt_eig_vecs, pt_eig_vals, fList);

 eos::morphablemodel::MorphableModel morphModel(shapeModel, colorModel, texCoords);
 eos::render::Mesh mesh = morphModel.get_mean();
 cout<< "vertices:" << mesh.vertices.size() << endl;

can you please share sample .scm model file (your internal model file with PCA data)

**Comment by samsgates** [comment]
Custom PCA model output:
vertices:0
pca count : 3

Morphable Model PCA output:
vertices :3448
pca count : 63

**Comment by patrikhuber** [comment]
Can you check the dimensions of your `pt_mean` that you put into the `PcaModel` constructor?

The .scm file has exactly the same content as the .bin file, you can inspect the member variables of `PcaModel`. See also #27.

**Comment by samsgates** [comment]
i am sure there is some mistake to calculate pt_mean,  how to create inputArray to place construct method in pt_pca  
cv::PCA pt_pca(**InputArray**, cv:Mat, CV_PCA_DATA_AS_ROW, 0);

i have two inputs from obj file 
1. vertices list  (845 lines - x y z) 
2. triangle list  (1610 lines - x y z) - not used to calculate PCA


======================== code ========
std::vector<std::array<int, 3>> triangleList;   //vertices list
    triangleList.resize(845);
    
     char* listFilename = argv[1];  //only vertices list from file (x y z, x y z, x y z....)
    int number_of_lines = 0;
    std::string line;
    std::ifstream listfile(listFilename);
    while (std::getline(listfile, line))
    {
        int xVal = 0;
        int yVal = 0;
        int zVal = 0;
        std::vector<std::string> splitstrs;
        boost::split(splitstrs, line, boost::is_any_of("\t "));
        xVal = stoi(splitstrs[0]);
        yVal = stoi(splitstrs[1]);
        zVal = stoi(splitstrs[2]);
        
        triangleList[number_of_lines][0] = xVal;
        triangleList[number_of_lines][1] = yVal;
        triangleList[number_of_lines][2] = zVal;
        
         //std::cout << xVal << ":" << yVal << ":" << zVal << endl;
        
        number_of_lines++;
    }
        
    int totallines=845; // total rows
    int feature_vector_size=3;  //total cols
    Mat input_feature_vector(totallines,feature_vector_size,CV_64FC1);
    
    for(int i=0;i<totallines;i++)
    {
        for(int j=0;j<feature_vector_size;j++)
        {
            input_feature_vector.at<double>(i,j)=triangleList[i][j];
            
        }
        
    }


cv::PCA pt_pca(input_feature_vector, cv::Mat(), CV_PCA_DATA_AS_ROW, 0);

cv::Mat pt_mean = pt_pca.mean;
cout<< "mean : " << pt_mean << endl;    
cv::Mat pt_eig_vals = pt_pca.eigenvalues;
cout<< "eigen val : " << pt_eig_vals << endl;
cv::Mat pt_eig_vecs = pt_pca.eigenvectors;

//========output=====
mean : [0, 0.3408284023668639, -32.9810650887574]
eigen val : [1913.358493803672;
 1284.189349112426;
 457.2918475709287]
eigen vector : [1.540067600050469e-16, 0.9997975675177745, -0.02012023820786627;
 1, -1.491034073135543e-16, 2.452026667564287e-16;
 -2.421530336991956e-16, 0.02012023820786627, 0.9997975675177745]
vertices:0
pca count : 3

**Comment by patrikhuber** [comment]
Your mean is just three values, so unless your model only has one vertex, something is wrong with it. You probably have to check the dimensions and content of your `input_feature_vector` variable.

**Comment by samsgates** [comment]
i will check with dimensions inputArray
one more question: Total vertices of Morphable model is 3448, but only 845 vertices are under mapping file (ibug2did.txt) - points mapped in landmark. 
if i add new PCA model how to edit this mapping file (68 point annotation mapped to vertices point)

**Comment by patrikhuber** [comment]
The `ibug2did.txt` file is for mapping 2D landmarks to 3D vertices. It defines the 68 (actually 66 I think) ibug points. See also the comment in the file:

`; Note: Points above vertex id 845 are not defined on the reference and thus not available in all model resolutions.`

You can of course define a mapping to *any* vertex, I just use the ones below id 845 because our models are hierarchical and it's great if a mapping is defined on *all* model resolution levels.

**Comment by tiancaipipi110** [comment]
@patrikhuber As I understand, the registered data is the aligned model scan, which could be done with a good 3d scanner or cleaned up by a modeler in couple minutes. Why is it the hard part?

**Comment by patrikhuber** [comment]
Hi @tiancaipipi110,

>> the registered data is the aligned model scan

Correct - aligned **and** brought into dense correspondence.

The hard part is *registering* the data, i.e. to bring it in dense correspondence. From a 3D scanner you will get raw data - that needs to be cleaned up ("easy", just time-consuming), but then you need a registration algorithm to bring that data in dense correspondence with each other / with a template. There's software for this available on the web as well, but from my experience I can tell you it's likely not something you can just run and it will work well on your data. Of course it always depends on what quality of results you are looking for.

**Comment by tiancaipipi110** [comment]
> bring that data in dense correspondence with each other / with a template. 
Could you please elaborate more on this? The cleaned up raw data you get from a 3d scanner of an i.e. apple, is already a complete 3d apple mesh, why do you need correspondence from this apple mesh to this apple mesh? Unless you meant correspondence from apple to the next scan of i.e. pear.

**Comment by tiancaipipi110** [comment]
@patrikhuber Or unless you meant using a very old scanner which only scans partial of an object each time, then we do need to put a series of these scans into correspondence before we can align them.

**Comment by ShaojunBian** [comment]
dense correspondence means after the registration for scans, the meshes have the same topology. For example, we have ten scanned female faces and ten scanned male faces, after the registration and cleaned up, we get ten female face meshes and ten male face meshes all in the same fine topology. These meshes are used to build PCA model.

**Comment by tiancaipipi110** [comment]
@ShaojunBian 

> ten scanned female faces

Ten scanned female faces as ten scans from ten females or ten scans of the same female?

**Comment by ShaojunBian** [comment]
from ten different females

**Comment by tiancaipipi110** [comment]
@ShaojunBian ok, this makes sense. And the algorithm used to build the dense correspondence is called ICP? Each mesh is one sample of Si = {x1,y1,z1,x2,y2,z2.....xm,ym,zm}^T,  and  we feed PCA with these samples to get a PCA model (also called the morphable model)? The dense correspondence means (x1,y1,z1) of S1 corresponds to (x1,y1,z1) of S2?

**Comment by patrikhuber** [comment]
@tiancaipipi110 Correct. There are many algorithms for dense correspondence, ICP is one of the most often used ones, yes. There are many variations and tweaks to ICP. You can find lots about that online / in literature, I would recommend that you look for a tutorial on basic statistical shape modelling.

**Comment by tiancaipipi110** [comment]
Cool, thanks guys. My clouds finally cleared up.

**Comment by tiancaipipi110** [comment]
@patrikhuber  In the PCA model M := (v¯,σ,V),  σ ∈ R^(n−1).  Is `σ` the standard deviation of eigenvectors V or the standard deviation of dataset? And why the morphable model is represented by eigenvalues and eigenvectors, not `σ` and eigenvectors ? Also why the number of principle components is n-1, not n?

**Comment by patrikhuber** [comment]
@tiancaipipi110 An introductory tutorial to PCA and/or statistical shape modelling should answer your questions?

**Comment by tiancaipipi110** [comment]
@patrikhuber Nah, I totally get it in PCA mathematically, but completely lost it in the code, since the shape of eigenvalues and standard deviation of pca components I get don't match.

**Comment by patrikhuber** [comment]
@tiancaipipi110 I think if you do a practical tutorial with PCA for shape modelling, e.g. in Matlab or Python, it will answer your questions. Like you can do a simple example with 2D coordinates ([x, y]) and 5 or 10 training examples (even manually created). One way to do it (which the tutorials will go through) is to create the mean-free data matrix (e.g. `A`), then compute the cov `A^tA`, then compute the eigenvectors and eigenvalues of this matrix and you're done. You can verify what you've done by generating new samples or/and projecting examples into that PCA space. You also have to watch out if/when/how you need to multiply (rescale) with the eigenvalues (i.e. in eos there's the `orthonormal basis` and `rescaled basis`, where in the latter, each principal component is already scaled with its respective eigenvalue). It takes some practice, take a sheet of paper, and as I suggested a simple 2D example and work through it line by line in Python/Matlab.

**Comment by tiancaipipi110** [comment]
@patrikhuber  Thanks. Yes, I'm trying in python on a `m x n` matix using `numpy.linalg.svd`. I'm not sure if I'm using the function wrong or what, the returned eigenvalues(s^2) has shape (m,), where I'm expecting (n,). 
```
cov = np.dot(datac, datac.T)
[u,s,v] = np.linalg.svd(conv) # returned s has shape(m,)
pcs = [np.dot(datac.T, item) for item in u.T ]
pcs = np.array([d / np.linalg.norm(d) for d in pcs])
```

**Comment by patrikhuber** [comment]
@tiancaipipi110 It depends of course if you arrange your data row- or col-wise, but perhaps you need to do `A^tA` instead of `AA^t`. One is the covariance matrix, one is the inner product matrix. If you arrange your data inside `A` row-wise, I think `A^tA` should be the covariance matrix.
Also don't overcomplicate things as a beginner, it looks like you might be using the inner product matrix and then post-multiplying the eigenvectors with `datac^t`. Just use the standard covariance matrix. It also helps checking the dimensions of each matrix on each line, and as I mentioned before, write it down on a piece of paper...

Perhaps you should follow a different tutorial or have a look at 2-3 other tutorials - if you get stuck in one tutorial or are not quite sure yet after doing one, just do one or two other ones. Different resources often highlight and focus on different things.

If you struggle further I'd recommend you posting on the StackExchange site (math or statistics section or something).

**Comment by tiancaipipi110** [comment]
@patrikhuber  My data is column wise, row-features,column-samples. I think it's this `[u,s,v] = np.linalg.svd(cov)`, it should have `np.linalg.svd(datac)`instead . Different tutorials are using this differently, don't know which one is correct.

**Comment by patrikhuber** [comment]
@tiancaipipi110 We might be using the terms row/col-wise differently. What I meant with row-wise is that each of your examples is one row of your data matrix. Some tutorials do this the other way round, which is fine too, it's just convention, and you just need to make sure you stick to a convention.

As far as I know you should definitely do PCA on the covariance matrix (or the inner-product matrix). It's probably a way too to do PCA directly on the data matrix, though that would give you a different outcome (if you want to know the difference, search about that specific question and/or post on one of the help sites I mentioned).
As I already said above, make sure you also work on mean-free data (subtract the mean from the data matrix).

**Comment by tiancaipipi110** [comment]
> What I meant with row-wise is that each of your examples is one row of your data matrix.
Yes I know, each of my examples is one column of data matrix, which means the covariance should be datac^T*datac right?
`datac `is already the centered m(row) by n(col) data matrix.

**Comment by patrikhuber** [comment]
I see. Each of your examples is a column of the data matrix.

Well I wrote above:

> If you arrange your data inside A row-wise, I think `A^tA` should be the covariance matrix.

This is also how it's done in the `eos::pca` code linked above.

Since your data is "col-wise", I think the covariance matrix would possibly be "the other way round", so `AA^t`. So not like you say in your last post.

I haven't checked this on paper so I might well be wrong, but I am also fairly certain that the implementation (and documentation) in `eos::pca` is correct. (However it's of course not impossible that I made a mistake!)

If you are still not sure, as recommended above, I'd suggest asking in a forum dedicated to such questions. And me and the people here would probably appreciate if you report back once you figure it out.
